### PR TITLE
fix: build atoms fail when inject

### PR DESCRIPTION
**Pull request by li-zhixin**

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -81,8 +81,8 @@ gulp.task('atoms:mkdir', function atomsMkdir () {
 
 gulp.task('atoms:inject', function atomsInject () {
   log('Injecting build file into Selenium build');
-  return fs.promises.copyFile(
```
Comment (mykola-mokhnach) [inline gulpfile.js:84]: what is the actual problem with this call?

File: package.json [hunk 2]
```diff
@@ -64,23 +64,24 @@
...
     "@appium/eslint-config-appium": "^6.0.0",
+    "@appium/gulp-plugins": "^7.0.0",
     "@semantic-release/git": "^10.0.1",
     "appium-ios-simulator": "^4.0.0",
     "chai": "^4.1.2",
     "chai-as-promised": "^7.1.1",
     "eslint-config-prettier": "^8.5.0",
     "fancy-log": "^2.0.0",
     "finalhandler": "^1.1.2",
+    "fs-extra": "^10.1.0",
```
Comment (mykola-mokhnach) [inline package.json:76]: do we really need to import an additional module to just copy a single file?

File: gulpfile.js (hunk 1 above)
Comment (li-zhixin) [reply gulpfile.js:84]: Here you actually expect the folder `atoms_build_dir` to be copied to `. /tmp/selenium/javascript/appium-atoms-driver, but instead the method to copy the file is being called.

Comment (mykola-mokhnach) [reply gulpfile.js:84]: thanks, I get it.

Usually I'm not very into adding additional imports when it is not really necessary. Would it be possible to add a custom implementation of async recursive folder copy?

Comment (li-zhixin) [reply gulpfile.js:84]: I see, how about now?

File: gulpfile.js [hunk 3]
```diff
@@ -23,6 +23,26 @@ const TEMP_BUILD_DIRECTORY_NAME = 'appium-atoms-driver';
 
 const ATOMS_BUILD_TARGET = 'build_atoms';
 
+const copyFolderRecursive = async (src, dest) => {
```
Comment (mykola-mokhnach) [inline gulpfile.js:26]: this could be a vanilla function. There is no point in making it anonymous

File: gulpfile.js [hunk 4]
```diff
@@ -23,6 +23,26 @@ const TEMP_BUILD_DIRECTORY_NAME = 'appium-atoms-driver';
...
+  const entries = await fs.promises.readdir(src, { withFileTypes: true });
+  try {
+    await fs.promises.access(dest, fs.constants.R_OK);
+  } catch (err) {
+    await fs.promises.mkdir(dest);
```
Comment (mykola-mokhnach) [inline gulpfile.js:31]: please add an option to make the dir creation recursive

File: gulpfile.js [hunk 5]
```diff
@@ -23,6 +23,26 @@ const TEMP_BUILD_DIRECTORY_NAME = 'appium-atoms-driver';
...
+  }
+  for (let entry of entries) {
```
Comment (mykola-mokhnach) [inline gulpfile.js:33]: let -> const

File: gulpfile.js [hunk 6]
```diff
@@ -23,6 +23,26 @@ const TEMP_BUILD_DIRECTORY_NAME = 'appium-atoms-driver';
...
+    const srcPath = path.join(src, entry.name);
+    const destPath = path.join(dest, entry.name);
+    if (entry.isDirectory()) {
+      await copyFolderRecursive(srcPath, destPath);
+    } else if (entry.isFile()) {
+      await fs.promises.copyFile(srcPath, destPath);
+    } else {
+      throw new Error(`Cannot copy: ${srcPath}`);
```
Comment (mykola-mokhnach) [inline gulpfile.js:41]: I'd say we simply skip an entry if it is not a file or directory

File: gulpfile.js [hunk 7]
```diff
@@ -81,8 +101,8 @@ gulp.task('atoms:mkdir', function atomsMkdir () {
 
 gulp.task('atoms:inject', function atomsInject () {
   log('Injecting build file into Selenium build');
-  return fs.promises.copyFile(
-    ATOMS_BUILD_DIRECTORY, `${SELENIUM_DIRECTORY}/javascript/${TEMP_BUILD_DIRECTORY_NAME}`
+  return copyFolderRecursive(
+      ATOMS_BUILD_DIRECTORY, `${SELENIUM_DIRECTORY}/javascript/${TEMP_BUILD_DIRECTORY_NAME}`
```
Comment (mykola-mokhnach) [inline gulpfile.js:105]: we could use `path.join` for the merge

**Review by mykola-mokhnach — APPROVED** [review]

**Comment by mykola-mokhnach** [comment]
Thanks @li-zhixin 
could you please sign CLA, so we can merge the PR?

**Comment by li-zhixin** [comment]
> Thanks @li-zhixin could you please sign CLA, so we can merge the PR?

done

File: gulpfile.js (hunk 7 above)
Comment (li-zhixin) [reply gulpfile.js:105]: Do you mean like this?
``` javascript
  return copyFolderRecursive(
    ATOMS_BUILD_DIRECTORY, path.join(SELENIUM_DIRECTORY, 'javascript', TEMP_BUILD_DIRECTORY_NAME)
  );
```

Comment (mykola-mokhnach) [reply gulpfile.js:105]: yes

**Comment by jlipps** [comment]
I am pretty sure our whole atoms build system is broken with the move in Selenium to Bazel?

**Pull request merged**